### Reading our data stored in parquet format

In [3]:
import pandas as pd
df = pd.read_parquet('final_data.parquet')

In [4]:
df.head()

Car_Id                 Status State_x  Model_Year           Make  \
0  844435               In Force      OK        1990  Manufacturer7   
1  844435               In Force      AL        1955  Manufacturer2   
2  844435  Customer Cancellation      OH        1973  Manufacturer2   
3  844435               In Force      OK        1990  Manufacturer7   
4  844435               In Force      AL        1955  Manufacturer2   

  Body_Style  Vehicle_Value  Annual_Miles_Driven  Business_Use  \
0     4 door        50000.0                   56             0   
1     4 door          500.0                14800             0   
2        SUV          500.0                 8439             0   
3     4 door        50000.0                   56             0   
4     4 door          500.0                14800             0   

   Antique_Vehicle  ...  Hh_Start_Date    Phone_Number   Zip_  State_y  \
0                1  ...     2022-11-18  (709) 379-9036  70442       OK   
1                0  ...     2022-11-18  (709) 379-9036  70442       OK   
2                0  ...     2022-11-18  (709) 379-9036  70442       OK   
3                1  ...     1995-01-26  (539) 357-6171  73146       AL   
4                0  ...     1995-01-26  (539) 357-6171  73146       AL   

   Country  Referral_Source  Date_Of_Birth  Marital_Status  Employment_Type  \
0      USA            Other     2002-11-06               D          Student   
1      USA            Other     2002-11-06               D          Student   
2      USA            Other     2002-11-06               D          Student   
3      USA           Friend     2035-08-01               M          Retired   
4      USA           Friend     2035-08-01               M          Retired   

  Income  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 29 columns]

### Checking columns before analysis

In [5]:
pd.set_option('display.max_columns', None)
print(df.head())

   Car_Id                 Status State_x  Model_Year           Make  \
0  844435               In Force      OK        1990  Manufacturer7   
1  844435               In Force      AL        1955  Manufacturer2   
2  844435  Customer Cancellation      OH        1973  Manufacturer2   
3  844435               In Force      OK        1990  Manufacturer7   
4  844435               In Force      AL        1955  Manufacturer2   

  Body_Style  Vehicle_Value  Annual_Miles_Driven  Business_Use  \
0     4 door        50000.0                   56             0   
1     4 door          500.0                14800             0   
2        SUV          500.0                 8439             0   
3     4 door        50000.0                   56             0   
4     4 door          500.0                14800             0   

   Antique_Vehicle  Lien  Lease  Driver_Safety_Discount  \
0                1     1      0                       0   
1                0     1      0                       0   

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 845805 entries, 0 to 845804
Data columns (total 29 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Car_Id                   845805 non-null  int64         
 1   Status                   845805 non-null  object        
 2   State_x                  845805 non-null  object        
 3   Model_Year               845805 non-null  int64         
 4   Make                     845805 non-null  object        
 5   Body_Style               845805 non-null  object        
 6   Vehicle_Value            845805 non-null  float64       
 7   Annual_Miles_Driven      845805 non-null  int64         
 8   Business_Use             845805 non-null  int64         
 9   Antique_Vehicle          845805 non-null  int64         
 10  Lien                     845805 non-null  int64         
 11  Lease                    845805 non-null  int64         
 12  Driver_Safety_Di

### Firstly, let's choose the active households

In [14]:
active_hh = df[df['Active_Hh'] == 1]

### 1. What is the average number of cars per household?

In [16]:
avg_cars_per_hh = active_hh.groupby('Hh_Id')['Car_Id'].count().mean()
print("Average number of cars per household are: ",avg_cars_per_hh)

Average number of cars per household are:  6.438854658172808


Looks like the average number of cars that each household has are around 6

### 2. How many cars are there by model year?

In [18]:
no_of_cars_modelyr = active_hh['Model_Year'].value_counts().sort_index()
print(no_of_cars_modelyr)

Model_Year
1952     3455
1953     3361
1954     3528
1955     3397
1956     3352
        ...  
2020    21909
2021    22289
2022    22041
2023    21979
2024    22195
Name: count, Length: 73, dtype: int64


Let's see the full output in the form of dictionary instead of the truncated output to better see the number of cars pers each year

In [26]:
cars_dict = no_of_cars_modelyr.to_dict()
for key, value in cars_dict.items():
    print(f"{key}: {value}")

1952: 3455
1953: 3361
1954: 3528
1955: 3397
1956: 3352
1957: 3354
1958: 3494
1959: 3378
1960: 3351
1961: 3354
1962: 3427
1963: 3335
1964: 3392
1965: 3371
1966: 3343
1967: 3462
1968: 3344
1969: 3375
1970: 3362
1971: 3479
1972: 3455
1973: 3340
1974: 3277
1975: 3447
1976: 3543
1977: 3393
1978: 3414
1979: 3444
1980: 3269
1981: 3477
1982: 3621
1983: 3349
1984: 3403
1985: 3336
1986: 3438
1987: 3326
1988: 3463
1989: 3487
1990: 3428
1991: 3377
1992: 3327
1993: 3382
1994: 3435
1995: 3386
1996: 3318
1997: 3421
1998: 3393
1999: 3451
2000: 3346
2001: 3416
2002: 21877
2003: 22033
2004: 22382
2005: 22023
2006: 22102
2007: 22295
2008: 22028
2009: 22425
2010: 21833
2011: 22317
2012: 21956
2013: 21941
2014: 21896
2015: 22142
2016: 22328
2017: 22386
2018: 21768
2019: 21990
2020: 21909
2021: 22289
2022: 22041
2023: 21979
2024: 22195


### 3. How many cars are there by make?

In [28]:
no_of_cars_make = active_hh['Make'].value_counts()
print(no_of_cars_make)

Make
Manufacturer2    135862
Manufacturer1    135577
Manufacturer3    135326
Manufacturer6     88643
Manufacturer7     88553
Manufacturer4     47160
Manufacturer5     47090
Name: count, dtype: int64


We can observe that Manufacturer2 has made the highest cars whereas Manufacturer5 made the least

### 4. Which cars are the safest?  What variables did you consider to define “safe”?

In [29]:
print(df['Driver_Safety_Discount'].unique()) # just checking if it has bool values

[0 1]


In [56]:
grouped_cars = active_hh.groupby('Car_Id').agg({
    'Annual_Miles_Driven': 'mean',
    'Vehicle_Safety_Discount': 'max',
    'Claim_Payout': 'max',
    'Make': 'first',
    'Model_Year': 'first'
}).reset_index()

#selecting cars with the chosen safety features and driven less than 100,000 miles
safe_cars = grouped_cars[
    (grouped_cars['Vehicle_Safety_Discount'] == 1) & 
    (grouped_cars['Claim_Payout'] == 0) &
    (grouped_cars['Annual_Miles_Driven'] < 100000)
]
#counting the occurrences of each make and model year
make_model_count = safe_cars.groupby(['Make', 'Model_Year']).size().reset_index(name='Count')
#sorting in descending order to get the most common makes and model years
best_make_model = make_model_count.sort_values(by='Count', ascending=False)
#picking the top 5 combinations of make and model year
top_5_make_model = best_make_model.head(5)
print(top_5_make_model)

              Make  Model_Year  Count
218  Manufacturer3        2024   1139
217  Manufacturer3        2023   1114
216  Manufacturer3        2022   1102
507  Manufacturer7        2021   1090
508  Manufacturer7        2022   1063


First, I checked the Vehicle Safety Discount. If a car has this discount, it means that insurance companies think it’s safe, which is a good sign.

Then I looked at Claim Payout. If a car has a zero payout, it means there haven't been any big accidents, so it seems like a safer car.

I also thought about how many miles the car has been driven each year. If a car has been driven less than 100,000 miles, it usually means it has been on the road less and has fewer chances to be in an accident.

By putting all these pieces together, I could find cars that are not just labeled as safe by the insurance companies but also haven’t had many accidents and haven’t been driven much.

After that, I wanted to see which makes and model years were the most common among these safer cars. This helps people know which brands or years are known for being safer, making it easier for them to choose a good car.

In the end, I found some recommendations for safe cars based on the data, which is super helpful for anyone looking to buy a vehicle. 

### 5. Which states have the largest households (defined as number of customers in a household)?

In [57]:
largest_households = active_hh.groupby('State_y')['Cust_Id'].nunique().sort_values(ascending=False)
print(largest_households)

State_y
KY    8469
AK    8421
SC    8415
AR    8335
MA    8320
AL    8319
MN    8311
WY    8285
WA    8282
NV    8274
CT    8267
MI    8234
HI    8210
NE    8190
GA    8180
SD    8176
NY    8168
NM    8150
IN    8106
MO    8099
MS    8055
RI    8038
OR    8032
TX    8010
AZ    8004
FL    7986
ND    7958
UT    7952
WV    7948
IL    7941
OH    7922
CA    7912
NH    7905
MT    7901
NC    7894
ME    7893
NJ    7871
VA    7856
WI    7849
IA    7834
MD    7828
KS    7823
OK    7793
DE    7791
ID    7774
LA    7750
VT    7705
PA    7647
TN    7588
CO    7524
Name: Cust_Id, dtype: int64


I believe 'State_x' column reflects where the car is registered or insured, so I used 'State_y' instead.

We can see that the state with largest households in our data is Knetucky with 8469 households.

### 6. How many active households are there as of 1/1/2021?

In [58]:
no_of_active_households = active_hh[active_hh['Hh_Start_Date'] <= '2021-01-01'].shape[0]
print("Number of active households as of 1/1/2021 are", no_of_active_households)

Number of active households as of 1/1/2021 are 621082


### 7. What is the average age of customers?

In [59]:
from datetime import datetime
active_hh['Age'] = (datetime.now() - active_hh['Date_Of_Birth']).dt.days // 365

average_age = active_hh['Age'].mean()
print("Average age of customers is ", average_age, "years")

Average age of customers is  6.694062762178732 years


/var/folders/pv/kll1prqs39jc2wvhvd31dfmr0000gn/T/ipykernel_7396/1051416248.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  active_hh['Age'] = (datetime.now() - active_hh['Date_Of_Birth']).dt.days // 365


The average age of customers appears to be around 6-7 years, which seems unusually low. This is likely due to incorrect dates of birth in the dataset, as some entries have future years like 2035, 2030 etc.

### 8. How much does age vary by region?

In [60]:
age_by_region = active_hh.groupby('State_y')['Age'].mean().sort_values()
print(age_by_region)

State_y
KS    5.810087
AZ    6.042755
ID    6.078821
MI    6.094298
CO    6.146912
VT    6.152435
IN    6.284193
MS    6.305649
HI    6.337557
NH    6.356517
AR    6.431744
ME    6.432321
NV    6.447848
OK    6.535502
PA    6.543306
RI    6.553080
NM    6.565708
NY    6.618732
MD    6.631240
TN    6.653019
MT    6.663015
WI    6.665013
FL    6.665637
OR    6.698507
UT    6.701130
NE    6.714899
CT    6.749694
AL    6.776795
NC    6.784102
IA    6.807485
MN    6.813659
AK    6.832922
SC    6.844500
DE    6.846598
VA    6.859173
ND    6.886365
IL    6.892871
KY    6.910237
MA    6.912776
OH    6.925658
TX    6.928003
LA    6.933995
WV    6.934750
GA    7.002594
WA    7.077952
SD    7.220001
NJ    7.230480
CA    7.261976
WY    7.411979
MO    7.591138
Name: Age, dtype: float64


Most states have an average age between 6 and 7 years(like the average age of customers found above). States like Kansas, Arizona, and Idaho have the lowest average ages, while Missouri and Wyoming have the highest. The consistently low average ages overall suggest that many customers may have birth dates inaccurately recorded as future years, which skews the true age distribution.

### 9.Which age group has the most expensive claims? 

In [61]:
#making age groups(since our data has an average age of 6, I will consider small age groups here
bins = [0, 5, 10, 15, 20, 25, 30]
labels = ['<5', '5-10', '10-15', '15-20', '20-25', '25-30']
active_hh['Age_Group'] = pd.cut(active_hh['Age'], bins=bins, labels=labels)

average_claim_by_age_group = active_hh.groupby('Age_Group')['Claim_Payout'].mean()
most_expensive_claims = average_claim_by_age_group.idxmax()
print("Age group with the most expensive claims is", most_expensive_claims)

Age group with the most expensive claims is 15-20


/var/folders/pv/kll1prqs39jc2wvhvd31dfmr0000gn/T/ipykernel_7396/4285617393.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  active_hh['Age_Group'] = pd.cut(active_hh['Age'], bins=bins, labels=labels)
/var/folders/pv/kll1prqs39jc2wvhvd31dfmr0000gn/T/ipykernel_7396/4285617393.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  average_claim_by_age_group = active_hh.groupby('Age_Group')['Claim_Payout'].mean()
